### PR TITLE
Add case verify_vdso.

### PR DESCRIPTION
**Pull request by LiliDeng**

### CODE REVIEW

File: microsoft/testsuites/core/timesync.py [hunk 1]
```diff
@@ -351,3 +352,33 @@ def timesync_chrony(self, node: Node) -> None:
...
+            2. Check output of 1, should not show the time related syscalls,
+             e.g. clock_gettime, gettimeofday
+        """,
+        priority=2,
+    )
+    def verify_vdso(self, node: Node) -> None:
+        time_pattern = re.compile("(time|gettimeofday|clock_gettime)", re.M)
+        strace = node.tools[Strace]
+        # 1. Run strace date 2>&1
+        output = strace.get("date 2>&1")
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: How about other vdso commands? We should cover as much as possible.

File: microsoft/testsuites/core/timesync.py [hunk 2]
```diff
@@ -351,3 +352,33 @@ def timesync_chrony(self, node: Node) -> None:
...
+        leading to performance bottleneck, Linux systems have a mechanism called vdso
+        which helps in gettimeoftheday and clock_gettime to be executed in userspace (no
+        syscall).
+
+        Steps:
+            1. Run strace date 2>&1
+            2. Check output of 1, should not show the time related syscalls,
+             e.g. clock_gettime, gettimeofday
+        """,
+        priority=2,
```
Comment (squirrelsc) [inline microsoft/testsuites/core/timesync.py]: It should be P1, if it doesn't spend more than 1 minute.

File: lisa/tools/strace.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+from lisa.executable import Tool
+from lisa.operating_system import Posix
+
+
+class Strace(Tool):
+    @property
+    def command(self) -> str:
+        return "strace"
+
+    def _check_exists(self) -> bool:
```
Comment (dsrivastavv) [inline lisa/tools/strace.py]: Do we need to override this? The default in parent class should handle this case

File: microsoft/testsuites/core/timesync.py [hunk 4]
```diff
@@ -351,3 +352,33 @@ def timesync_chrony(self, node: Node) -> None:
         chrony.check_sources_and_stats()
         # 5. Check chrony tracking.
         chrony.check_tracking()
+
+    @TestCaseMetadata(
+        description="""
+        This test is to check gettime calls are not being redirected as system calls,
+        leading to performance bottleneck, Linux systems have a mechanism called vdso
```
Comment (dsrivastavv) [inline microsoft/testsuites/core/timesync.py]: [nit] bottleneck.

File: microsoft/testsuites/core/timesync.py [hunk 5]
```diff
@@ -351,3 +352,33 @@ def timesync_chrony(self, node: Node) -> None:
...
+        which helps in gettimeoftheday and clock_gettime to be executed in userspace (no
+        syscall).
+
+        Steps:
+            1. Run strace date 2>&1
+            2. Check output of 1, should not show the time related syscalls,
+             e.g. clock_gettime, gettimeofday
```
Comment (dsrivastavv) [inline microsoft/testsuites/core/timesync.py]: Is the following exhaustive: time, gettimeofday, clock_gettime ? If so we can mention these directly here

File: microsoft/testsuites/core/timesync.py [hunk 6]
```diff
@@ -351,3 +352,33 @@ def timesync_chrony(self, node: Node) -> None:
...
+        """,
+        priority=2,
+    )
+    def verify_vdso(self, node: Node) -> None:
+        time_pattern = re.compile("(time|gettimeofday|clock_gettime)", re.M)
```
Comment (dsrivastavv) [inline microsoft/testsuites/core/timesync.py]: [nit] can be a global variable in the class

File: microsoft/testsuites/core/timesync.py (hunk 1 above)
Comment (LiliDeng) [reply microsoft/testsuites/core/timesync.py]: - https://github.com/nlynch-mentor/vdsotest
- https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/tree/tools/testing/selftests/vDSO

I found 2 repos which can test vdso. After trying, found the second one can't repro the bug. The first one is easy to compile and can repro the issue, also it contains other vdso commands, do you think it is feasible to use it?

Comment (squirrelsc) [reply microsoft/testsuites/core/timesync.py]: ok, we can use the first one. And add your explanation in suite description to explain why we don't use the kernel tests.

Comment (LiliDeng) [reply microsoft/testsuites/core/timesync.py]: update

File: microsoft/testsuites/core/timesync.py (hunk 2 above)
Comment (LiliDeng) [reply microsoft/testsuites/core/timesync.py]: update

File: lisa/tools/strace.py (hunk 3 above)
Comment (LiliDeng) [reply lisa/tools/strace.py]: remove

File: microsoft/testsuites/core/timesync.py (hunk 4 above)
Comment (LiliDeng) [reply microsoft/testsuites/core/timesync.py]: update

File: microsoft/testsuites/core/timesync.py (hunk 5 above)
Comment (LiliDeng) [reply microsoft/testsuites/core/timesync.py]: now change to run vdso benchmark, this code was removed

File: microsoft/testsuites/core/timesync.py (hunk 6 above)
Comment (LiliDeng) [reply microsoft/testsuites/core/timesync.py]: now change to run vdso benchmark, this code was removed

**Review by squirrelsc — APPROVED** [review]

**Pull request merged**